origin/blobclient/cluster_client.go

```diff
@@ -123,23 +123,41 @@ func (c *clusterClient) CheckReadiness() error {
 }
 
 // UploadBlob uploads blob to origin cluster. See Client.UploadBlob for more details.
-func (c *clusterClient) UploadBlob(namespace string, d core.Digest, blob io.Reader) (err error) {
+func (c *clusterClient) UploadBlob(namespace string, d core.Digest, blob io.ReadSeeker) (err error) {
 	clients, err := c.resolver.Resolve(d)
 	if err != nil {
 		return fmt.Errorf("resolve clients: %s", err)
 	}
+	logger := log.With("namespace", namespace, "digest", d.Hex())
+	logger.Debug("Starting blob upload to origin cluster")
 
 	// We prefer the origin with highest hashing score so the first origin will handle
 	// replication to origins with lower score. This is because we want to reduce upload
 	// conflicts between local replicas.
-	for _, client := range clients {
+	for i, client := range clients {
+		originAddr := client.Addr()
+		attemptLogger := logger.With("origin", originAddr, "attempt", i)
+
+		attemptLogger.Debug("Attempting blob upload to origin")
 		err = client.UploadBlob(namespace, d, blob)
+		if err == nil {
+			attemptLogger.Debug("Blob upload succeeded")
+			return nil
+		}
+		attemptLogger.With("error", err).Error("Blob upload failed")
+
+		// Non-retryable error - don't try other origins
+		if !httputil.IsNetworkError(err) && !httputil.IsRetryable(err) {
+			break
```

Comment:
```suggestion
			return err
```

nit: `return err` makes it easier for the reader to understand what's happening n this case, compared to `break`, which then requires the reader to go through the rest of the code to see what the final result is.